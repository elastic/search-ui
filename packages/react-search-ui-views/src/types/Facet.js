import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
