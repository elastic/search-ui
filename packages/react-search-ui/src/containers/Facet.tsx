import React from "react";
import { Component } from "react";
import {
  FacetContainerContext,
  FacetContainerProps,
  FacetViewProps,
  MultiCheckboxFacet
} from "@elastic/react-search-ui-views";
import { helpers } from "@elastic/search-ui";

import { accentFold } from "../helpers";
import { withSearch } from "..";

const { markSelectedFacetValuesFromFilters } = helpers;

type FacetContainerState = {
  searchTerm: string;
  more: number;
};

export class FacetContainer extends Component<
  FacetContainerProps,
  FacetContainerState
> {
  static defaultProps = {
    filterType: "all",
    isFilterable: false,
    show: 5
  };

  constructor(props) {
    super(props);
    this.state = {
      more: props.show,
      searchTerm: ""
    };
  }

  handleClickMore = (totalOptions) => {
    this.setState(({ more }) => {
      let visibleOptionsCount = more + 10;
      const showingAll = visibleOptionsCount >= totalOptions;
      if (showingAll) visibleOptionsCount = totalOptions;

      this.props.a11yNotify("moreFilters", { visibleOptionsCount, showingAll });

      return { more: visibleOptionsCount };
    });
  };

  handleFacetSearch = (searchTerm) => {
    this.setState({ searchTerm });
  };

  render() {
    const { more, searchTerm } = this.state;
    const {
      addFilter,
      className,
      facets,
      field,
      filterType,
      filters,
      label,
      removeFilter,
      setFilter,
      view,
      isFilterable,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      a11yNotify,
      ...rest
    } = this.props;
    const facetsForField = facets[field];

    if (!facetsForField) return null;

    // By using `[0]`, we are currently assuming only 1 facet per field. This will likely be enforced
    // in future version, so instead of an array, there will only be one facet allowed per field.
    const facet = facetsForField[0];

    let facetValues = markSelectedFacetValuesFromFilters(
      facet,
      filters,
      field,
      filterType
    ).data;

    const selectedValues = facetValues
      .filter((fv) => fv.selected)
      .map((fv) => fv.value);

    if (!facetValues.length && !selectedValues.length) return null;

    if (searchTerm.trim()) {
      facetValues = facetValues.filter((option) => {
        let valueToSearch;
        switch (typeof option.value) {
          case "string":
            valueToSearch = accentFold(option.value).toLowerCase();
            break;
          case "number":
            valueToSearch = option.value.toString();
            break;
          case "object":
            valueToSearch =
              typeof option?.value?.name === "string"
                ? accentFold(option.value.name).toLowerCase()
                : "";
            break;

          default:
            valueToSearch = "";
            break;
        }
        return valueToSearch.includes(accentFold(searchTerm).toLowerCase());
      });
    }

    const View: React.ComponentType<FacetViewProps> =
      view || MultiCheckboxFacet;

    const viewProps: FacetViewProps = {
      className,
      label: label,
      onMoreClick: this.handleClickMore.bind(this, facetValues.length),
      onRemove: (value) => {
        removeFilter(field, value, filterType);
      },
      onChange: (value) => {
        setFilter(field, value, filterType);
      },
      onSelect: (value) => {
        addFilter(field, value, filterType);
      },
      options: facetValues.slice(0, more),
      showMore: facetValues.length > more,
      values: selectedValues,
      showSearch: isFilterable,
      onSearch: (value) => {
        this.handleFacetSearch(value);
      },
      searchPlaceholder: `Filter ${label}`,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<FacetContainerProps, FacetContainerContext>(
  ({ filters, facets, addFilter, removeFilter, setFilter, a11yNotify }) => ({
    filters,
    facets,
    addFilter,
    removeFilter,
    setFilter,
    a11yNotify
  })
)(FacetContainer);
