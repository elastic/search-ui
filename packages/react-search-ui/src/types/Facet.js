import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  // Name of the field this facet is associated with
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
