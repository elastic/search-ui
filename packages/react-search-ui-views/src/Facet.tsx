import {
  FacetValue,
  FieldValue,
  FilterType,
  SearchContextState
} from "@elastic/search-ui";
import { BaseContainerProps } from "./types";

export type FacetViewProps = {
  className?: string;
  label: string;
  onMoreClick: () => void;
  onRemove: (value: FieldValue) => void;
  onChange: (value: FieldValue) => void;
  onSelect: (value: FieldValue) => void;
  options: FacetValue[];
  showMore: boolean;
  values: any;
  showSearch: boolean;
  onSearch: (value: string) => void;
  searchPlaceholder: string;
};

export type FacetContainerContext = Pick<
  SearchContextState,
  | "addFilter"
  | "removeFilter"
  | "setFilter"
  | "a11yNotify"
  | "filters"
  | "facets"
>;

export type FacetContainerProps = BaseContainerProps & {
  filterType: FilterType;
  show?: number;
  view?: React.ComponentType<FacetViewProps>;
  isFilterable: boolean;
  field: string;
  label: string;
} & FacetContainerContext;
