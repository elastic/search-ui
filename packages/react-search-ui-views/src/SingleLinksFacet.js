import PropTypes from "prop-types";
import React from "react";

import { FacetValue } from "./types";
import { getFilterValueDisplay } from "./view-helpers";
import { appendClassName } from "./view-helpers";

function SingleLinksFacet({ className, label, onRemove, onSelect, options }) {
  const value = options.filter(o => o.selected).map(o => o.value)[0];
  return (
    <div className={appendClassName("sui-facet", className)}>
      <div>
        <div className="sui-facet__title">{label}</div>
        <ul className="sui-single-option-facet">
          {value && (
            <li className="sui-single-option-facet__selected">
              {getFilterValueDisplay(value)}{" "}
              <span className="sui-single-option-facet__remove">
                (
                <a
                  onClick={e => {
                    e.preventDefault();
                    onRemove(value);
                  }}
                  href="/"
                >
                  Remove
                </a>
                )
              </span>
            </li>
          )}
          {!value &&
            options.map(option => (
              <li
                className="sui-single-option-facet__item"
                key={getFilterValueDisplay(option.value)}
              >
                <a
                  className="sui-single-option-facet__link"
                  href="/"
                  onClick={e => {
                    e.preventDefault();
                    onSelect(option.value);
                  }}
                >
                  {getFilterValueDisplay(option.value)}
                </a>{" "}
                <span className="sui-facet__count">{option.count}</span>
              </li>
            ))}
        </ul>
      </div>
    </div>
  );
}

SingleLinksFacet.propTypes = {
  label: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetValue).isRequired,
  className: PropTypes.string
};

export default SingleLinksFacet;
