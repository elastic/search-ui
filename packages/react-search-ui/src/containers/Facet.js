import PropTypes from "prop-types";
import { Component } from "react";
import { MultiCheckboxFacet } from "@elastic/react-search-ui-views";
import { helpers } from "@elastic/search-ui";

import { Facet, Filter, FilterType } from "../types";
import { accentFold } from "../helpers";
import { withSearch } from "..";

const { markSelectedFacetValuesFromFilters } = helpers;

export class FacetContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    field: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    filterType: FilterType,
    show: PropTypes.number,
    view: PropTypes.func,
    isFilterable: PropTypes.bool,
    // State
    filters: PropTypes.arrayOf(Filter).isRequired,
    facets: PropTypes.objectOf(PropTypes.arrayOf(Facet)).isRequired,
    // Actions
    addFilter: PropTypes.func.isRequired,
    removeFilter: PropTypes.func.isRequired,
    setFilter: PropTypes.func.isRequired,
    a11yNotify: PropTypes.func.isRequired
  };

  static defaultProps = {
    filterType: "all",
    isFilterable: false
  };

  constructor({ show = 5 }) {
    super();
    this.state = {
      more: show,
      searchTerm: ""
    };
  }

  handleClickMore = totalOptions => {
    this.setState(({ more }) => {
      let visibleOptionsCount = more + 10;
      const showingAll = visibleOptionsCount >= totalOptions;
      if (showingAll) visibleOptionsCount = totalOptions;

      this.props.a11yNotify("moreFilters", { visibleOptionsCount, showingAll });

      return { more: visibleOptionsCount };
    });
  };

  handleFacetSearch = searchTerm => {
    this.setState({ searchTerm });
  };

  render() {
    const { more, searchTerm } = this.state;
    const {
      addFilter,
      className,
      facets,
      field,
      filterType,
      filters,
      label,
      removeFilter,
      setFilter,
      view,
      isFilterable,
      // eslint-disable-next-line no-unused-vars
      a11yNotify,
      ...rest
    } = this.props;
    const facetsForField = facets[field];

    if (!facetsForField) return null;

    // By using `[0]`, we are currently assuming only 1 facet per field. This will likely be enforced
    // in future version, so instead of an array, there will only be one facet allowed per field.
    const facet = facetsForField[0];

    let facetValues = markSelectedFacetValuesFromFilters(
      facet,
      filters,
      field,
      filterType
    ).data;

    const selectedValues = facetValues
      .filter(fv => fv.selected)
      .map(fv => fv.value);

    if (!facetValues.length && !selectedValues.length) return null;

    if (searchTerm.trim()) {
      facetValues = facetValues.filter(option =>
        accentFold(option.value)
          .toLowerCase()
          .includes(accentFold(searchTerm).toLowerCase())
      );
    }

    const View = view || MultiCheckboxFacet;

    return View({
      className,
      label: label,
      onMoreClick: this.handleClickMore.bind(this, facetValues.length),
      onRemove: value => {
        removeFilter(field, value, filterType);
      },
      onChange: value => {
        setFilter(field, value, filterType);
      },
      onSelect: value => {
        addFilter(field, value, filterType);
      },
      options: facetValues.slice(0, more),
      showMore: facetValues.length > more,
      values: selectedValues,
      showSearch: isFilterable,
      onSearch: value => {
        this.handleFacetSearch(value);
      },
      searchPlaceholder: `Filter ${field}`,
      ...rest
    });
  }
}

export default withSearch(
  ({ filters, facets, addFilter, removeFilter, setFilter, a11yNotify }) => ({
    filters,
    facets,
    addFilter,
    removeFilter,
    setFilter,
    a11yNotify
  })
)(FacetContainer);
