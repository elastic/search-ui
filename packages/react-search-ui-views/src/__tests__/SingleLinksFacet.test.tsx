import React from "react";
import SingleLinksFacet from "../SingleLinksFacet";
import { shallow } from "enzyme";
import { FacetViewProps } from "../Facet";

const params: FacetViewProps = {
  label: "Facet",
  onRemove: jest.fn(),
  onSelect: jest.fn(),
  options: [
    { value: "1", count: 1, selected: false },
    { value: "2", count: 1, selected: false }
  ],
  onChange: jest.fn(),
  onMoreClick: jest.fn(),
  onSearch: jest.fn(),
  searchPlaceholder: "Search",
  showMore: false,
  showSearch: false,
  values: []
};

it("renders correctly", () => {
  const wrapper = shallow(<SingleLinksFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders falsey values correctly", () => {
  const wrapper = shallow(
    <SingleLinksFacet
      {...params}
      options={[
        {
          value: 0,
          count: 10,
          selected: false
        },
        {
          value: false,
          count: 20,
          selected: false
        },
        {
          value: "",
          count: 30,
          selected: false
        }
      ]}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

describe("determining selected option", () => {
  it("will correctly determine which of the options is selected", () => {
    const wrapper = shallow(
      <SingleLinksFacet
        {...params}
        options={[
          { value: "1", count: 1, selected: true },
          { value: "2", count: 1, selected: false }
        ]}
      />
    );
    expect(wrapper.find("li").length).toBe(1);
    expect(wrapper.find("li").text()).toBe("1 (Remove)");
  });

  // This shouldn't ever happen, but if it does, it should use the first selected value
  it("will used the first selected option when multiple options are selected", () => {
    const wrapper = shallow(
      <SingleLinksFacet
        {...params}
        options={[
          { value: "1", count: 1, selected: true },
          { value: "2", count: 1, selected: true }
        ]}
      />
    );
    expect(wrapper.find("li").length).toBe(1);
    expect(wrapper.find("li").text()).toBe("1 (Remove)");
  });

  it("will correctly determine when no value is selected", () => {
    const wrapper = shallow(
      <SingleLinksFacet
        {...params}
        options={[
          { value: "1", count: 1, selected: false },
          { value: "2", count: 1, selected: false }
        ]}
      />
    );
    expect(wrapper.find("li").length).toBe(2);
  });
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <SingleLinksFacet {...params} className={customClassName} />
  );
  const { className } = wrapper.props();
  expect(className).toEqual("sui-facet test-class");
});
