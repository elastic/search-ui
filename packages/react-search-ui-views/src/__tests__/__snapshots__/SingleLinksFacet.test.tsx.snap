// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sui-facet"
>
  <div>
    <div
      className="sui-facet__title"
    >
      Facet
    </div>
    <ul
      className="sui-single-option-facet"
    >
      <li
        className="sui-single-option-facet__item"
        key="1"
      >
        <a
          className="sui-single-option-facet__link"
          href="/"
          onClick={[Function]}
        >
          1
        </a>
         
        <span
          className="sui-facet__count"
        >
          1
        </span>
      </li>
      <li
        className="sui-single-option-facet__item"
        key="2"
      >
        <a
          className="sui-single-option-facet__link"
          href="/"
          onClick={[Function]}
        >
          2
        </a>
         
        <span
          className="sui-facet__count"
        >
          1
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders falsey values correctly 1`] = `
<div
  className="sui-facet"
>
  <div>
    <div
      className="sui-facet__title"
    >
      Facet
    </div>
    <ul
      className="sui-single-option-facet"
    >
      <li
        className="sui-single-option-facet__item"
        key="0"
      >
        <a
          className="sui-single-option-facet__link"
          href="/"
          onClick={[Function]}
        >
          0
        </a>
         
        <span
          className="sui-facet__count"
        >
          10
        </span>
      </li>
      <li
        className="sui-single-option-facet__item"
        key="false"
      >
        <a
          className="sui-single-option-facet__link"
          href="/"
          onClick={[Function]}
        >
          false
        </a>
         
        <span
          className="sui-facet__count"
        >
          20
        </span>
      </li>
      <li
        className="sui-single-option-facet__item"
        key=""
      >
        <a
          className="sui-single-option-facet__link"
          href="/"
          onClick={[Function]}
        />
         
        <span
          className="sui-facet__count"
        >
          30
        </span>
      </li>
    </ul>
  </div>
</div>
`;
